src/history/base.rs

```diff
@@ -14,42 +20,285 @@ pub enum HistoryNavigationQuery {
+        SearchFilter {
+            command_line: None,
+            not_command_line: None,
+            hostname: None,
+            cwd_exact: None,
+            cwd_prefix: None,
+            exit_successful: None,
+        }
+    }
+}
+
+pub struct SearchQuery {
+    pub direction: SearchDirection,
+    /// if given, only get results after/before this time (depending on direction)
+    pub start_time: Option<chrono::DateTime<Utc>>,
+    pub end_time: Option<chrono::DateTime<Utc>>,
+    /// if given, only get results after/before this id (depending on direction)
+    pub start_id: Option<HistoryItemId>,
+    pub end_id: Option<HistoryItemId>,
+    pub limit: Option<i64>,
+    pub filter: SearchFilter,
+}
+/// some utility functions
+impl SearchQuery {
+    /// all that contain string in reverse chronological order
+    pub fn all_that_contain_rev(contains: String) -> SearchQuery {
+        SearchQuery {
+            direction: SearchDirection::Backward,
+            start_time: None,
+            end_time: None,
+            start_id: None,
+            end_id: None,
+            limit: None,
+            filter: SearchFilter::from_text_search(CommandLineSearch::Substring(contains)),
+        }
+    }
+    pub fn last_with_search(filter: SearchFilter) -> SearchQuery {
+        SearchQuery {
+            direction: SearchDirection::Backward,
+            start_time: None,
+            end_time: None,
+            start_id: None,
+            end_id: None,
+            limit: Some(1),
+            filter,
+        }
+    }
+    pub fn last_with_prefix(prefix: String) -> SearchQuery {
+        SearchQuery::last_with_search(SearchFilter::from_text_search(CommandLineSearch::Prefix(
+            prefix,
+        )))
+    }
+    pub fn everything(direction: SearchDirection) -> SearchQuery {
+        SearchQuery {
+            direction,
+            start_time: None,
+            end_time: None,
+            start_id: None,
+            end_id: None,
+            limit: None,
+            filter: SearchFilter::anything(),
+        }
+    }
+}
+
+/// Represents a history file or database
+/// Data could be stored e.g. in a plain text file, in a JSONL file, in a SQLite database
 pub trait History: Send {
-    /// Append entry to the history, if capacity management is part of the implementation may perform that as well
-    fn append(&mut self, entry: &str);
+    /// save a history item to the database
+    /// if given id is None, a new id is created and set in the return value
+    /// if given id is Some, the existing entry is updated
+    fn save(&mut self, h: HistoryItem) -> Result<HistoryItem>;
+    /// load a history item by its id
+    fn load(&self, id: HistoryItemId) -> Result<HistoryItem>;
+    /// gets the newest item id in this history
+    fn newest(&self) -> Result<Option<HistoryItemId>> {
+        let res = self.search(SearchQuery::last_with_search(SearchFilter::anything()))?;
+        Ok(res.get(0).and_then(|e| e.id))
```

Comment:
Caveat: Backseat coding opinion

I would love to have potential IDs hidden as much as possible from regular outside use, as they will fail ops on `FilebackedHistory`

Reply:
Actually this function is not used at all anymore, I only needed it to construct a new HistoryCursor correctly but now I can remove it